### DISCUSSION TITLE
Could not load cimgui.so Linux

### DISCUSSION THREAD

**Discussion by pihentagy** [cavariux/nimgl#58] [Q&A] [answered]
* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] support request => Please do not submit support request here, see note
  at the top of this template.

* **What is the current behavior?**

The example does not compile. After fixing the obvious issue it does not find `cimgui.so`.

Please advice how to obtain the shared object.

* **If the current behavior is a bug, please provide the steps to reproduce and
if possible a minimal demo of the problem**

```
~ ⌚ 20:56:24
$ nim r i.nim                                      
Hint: used config file '/home/pihentagy/.choosenim/toolchains/nim-1.4.0/config/nim.cfg' [Conf]
Hint: used config file '/home/pihentagy/.choosenim/toolchains/nim-1.4.0/config/config.nims' [Conf]
............CC: stdlib_system.nim
CC: .nimble/pkgs/imgui-1.78.0/imgui.nim
CC: i.nim

Hint:  [Link]
Hint: 36943 lines; 1.612s; 60.09MiB peakmem; Debug build; proj: /home/pihentagy/i.nim; out: /home/pihentagy/.cache/nim/i_d/i [SuccessX]
Hint: /home/pihentagy/.cache/nim/i_d/i  [Exec]
could not load: cimgui.so
(compile with -d:nimDebugDlOpen for more information)
Error: execution of an external program failed: '/home/pihentagy/.cache/nim/i_d/i '

~ ⌚ 20:56:37
$ 
```

The corrected example code
```nim
import imgui

igText("Hello World")

var f: ptr float32
igSliderFloat("float", f, 0.0f, 1.0f)
```

* **Please tell us about your environment:**
  
  - Version: nimgl@1.1.5, imgui@1.78
  - OS: Linux
  - Bindings: dunno how can I choose, could you point me to a readme or add relevant info to the readme?

**Comment by lmariscal** [comment, marked as answer]
Since ImGui is a C++ project to interact with it from the C backend we need a DLL. To generate a DLL I recommend using CMake and the official fork of cimgui https://github.com/nimgl/cimgui.

```bash
git clone --recursive https://github.com/nimgl/cimgui.git
cd cimgui
mkdir bld
cd bld
cmake ..
make # In case you are using Makefiles
```

If you followed the previous commands, there should be `cimgui.so` file, copy that file to your executable directory.

Please let me know if you have any other doubts.

**Reply by pihentagy** [reply]
Ok, fine, the `.so` file gets created. (I'm on linux (mint), so with help of make I got `cimgui.so`.

I also did `nimble install nimgl`.

```console
/tmp/cimgui on  master! ⌚ 14:09:42
$ cat main.nim      
import imgui

igText("Hello World")

var f: ptr float32
igSliderFloat("float", f, 0.0f, 1.0f)

/tmp/cimgui on  master! ⌚ 14:10:51
$ nim c main.nim   
Hint: used config file '/home/pihentagy/.choosenim/toolchains/nim-1.4.0/config/nim.cfg' [Conf]
Hint: used config file '/home/pihentagy/.choosenim/toolchains/nim-1.4.0/config/config.nims' [Conf]
............
Hint:  [Link]
Hint: 36943 lines; 0.616s; 49.219MiB peakmem; Debug build; proj: /tmp/cimgui/main.nim; out: /tmp/cimgui/main [SuccessX]

/tmp/cimgui on  master! ⌚ 14:10:58
$ ./main 
Traceback (most recent call last)
/tmp/cimgui/main.nim(3)  main
SIGSEGV: Illegal storage access. (Attempt to read from nil?)


```

**Reply by lmariscal** [reply]
The issue here is that you are trying to write ImGui data without a context created. Please check the provided example in the examples folder in the main repo: https://github.com/nimgl/nimgl/blob/master/examples/timgui.nim or one of the tests from the test folder in the imgui subrepo https://github.com/nimgl/imgui/tree/master/tests.
To simply verify that it is working I would recommend using the tnull.nim https://github.com/nimgl/imgui/blob/master/tests/tnull.nim since you don't have to create a visual context to check ImGui bindings.
